Remove unnecessary remote exchange from scalar correlated subquery plan

mbasmanova opened the pull request:
Fixes #10830

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java:
```diff
@@ -0,0 +1,119 @@
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class PruneMarkDistinctSymbols
```
Call this PruneMarkDistinctColumns (It's what I've been using for all other PruneXXX rules)

martint replied:
I'm not sure I understand what this rule is doing. Can you add an explanation to the commit message or illustrate what the original vs transformed plan looks like? (in the javadoc of this class, for instance, like we do in other rules)

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java:
```diff
@@ -0,0 +1,119 @@
+    @Override
+    public Result apply(MarkDistinctNode node, Captures captures, Context context)
+    {
+        List<LocalProperty<Symbol>> localProperties = derivePropertiesRecursively(node.getSource(), context.getLookup(), metadata, context.getSession(), context.getSymbolAllocator().getTypes(), sqlParser)
```
Please add a comment to indicate this is a temporary hack and not to be emulated in other rules without careful consideration

mbasmanova commented:
@martint @sopel39 Martin, I updated commit message and added comments to the new rule. Would you take another look?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java:
```diff
@@ -0,0 +1,141 @@
+        // Check if all non-trailing constant symbols are included in node.getDistinctSymbols().
+        // If so, the trailing constant symbols can be dropped from node.getDistinctSymbols().
+
+        if (!node.getDistinctSymbols().containsAll(prefixBuilder.build())) {
```
Maybe add comment that all prefix symbols are implicitly `grouped`? This is definitely not obvious from the code and requires knowledge how local properties are represented and what they mean.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java:
```diff
@@ -0,0 +1,141 @@
+                prefixBuilder.addAll(property.getColumns());
+            }
+            else {
+                suffixBuilder.addAll(property.getColumns());
```
Maybe rename `suffixBuilder` to `constantSymbols`?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            if (foundNonConstant) {
+                prefixBuilder.addAll(property.getColumns());
```
Maybe rename `prefixBuilder` to `groupedSymbols`?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctSymbols.java:
```diff
@@ -0,0 +1,141 @@
+        for (int i = localProperties.size() - 1; i >= 0; i--) {
+            LocalProperty<Symbol> property = localProperties.get(i);
+            if (!foundNonConstant && !(property instanceof ConstantProperty)) {
+                foundNonConstant = true;
```
you could break loop here and just take prefix as a sublist, or just make a loop that searches for split index:
```
for (splitIndex = localProperties.size(); splitIndex > 0; splitIndex--) {
  if (!(localProperties.get(splitIndex - 1) instanceof ConstantProperty)) {
    break;
  }
}
groupedColums = localProperties.subList(0, i+1).stream()
  .flatMap(LocalProperty::getColumns)
  .collect(toImmutableSet);
constantColumns = localProperties.subList(i+1, localProperties.size()).stream()
  .flatMap(LocalProperty::getColumns)
  .collect(toImmutableSet);
```

mbasmanova commented:
@sopel39 Karol, thank you for review. I'll be on PTO til July 29, but I'll address your comments once I'm back. Martin also asked to move the logic from the new `PruneMarkDistinctSymbols` rule to `AddLocalExchanges`. I'll do that as well.

mbasmanova commented:
@martint @sopel39 Karol, Martin, I addressed your comments and simplified the logic using functional dependencies.

mbasmanova commented:
@sopel39 Karol, I noticed that `TestTpcdsLocalStats.{testCharComparison, testDecimalComparison}` started failing. It turns out that sometimes we have redundant `MarkDistinctNode` instance where all `distinctSymbols` as constant. E.g. in the following query:

`SELECT count(*),count(distinct cd_marital_status) FROM (SELECT * FROM customer_demographics WHERE cd_marital_status = 'D')`

In this case, the new pruning logic was dropping all distinct symbols and `HashGenerationOptimizer.Rewriter#visitMarkDistinct` hit an `NPE` trying to add a hash symbol.

Long term I think we should optimize such queries to replace `MarkDistinctNode` with `Project( is_distinct := true)`, but for this PR I updated `AddLocalExchanges.Rewriter#pruneMarkDistinctSymbols` to keep at least one distinct symbol. I also updated `MarkDistinctNode` to fail on `distinctSymbols.empty()`.

Would you take another look before I merge?

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -366,8 +367,72 @@ public PlanWithProperties visitWindow(WindowNode node, StreamPreferredProperties
+                }
+            }
+
+            ImmutableSet<Symbol> redundantSymbols = redundantSymbolsBuilder.build();
```
Use `Set` for the type declaration.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -366,8 +367,72 @@ public PlanWithProperties visitWindow(WindowNode node, StreamPreferredProperties
+                return node.getDistinctSymbols();
+            }
+
+            // In case when all distinct symbols are constants, keep the first symbol (don't drop them all).
```
Add a TODO explaining that this is a workaround and requires a longer term fix.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -366,8 +367,72 @@ public PlanWithProperties visitWindow(WindowNode node, StreamPreferredProperties
+            }
+
+            // In case when all distinct symbols are constants, keep the first symbol (don't drop them all).
+            if (localProperties.get(0) instanceof ConstantProperty && (node.getDistinctSymbols().contains(((ConstantProperty<Symbol>) localProperties.get(0)).getColumn()))) {
```
It's not obvious how this relates to "in case all distinct symbols are constants...". It seems to be checking a single column (?)

mbasmanova commented:
@martint Martin, thanks for taking a look. As you pointed out, the logic for handing all-constant-distinct-symbols was wrong. I fixed it to (1) prune distinct symbols; (2) if no symbols left, return one of the original distinct symbols. 

Re: workaround - I take it back. We can't replace `MarkDistinct` with all-constant distinct symbols with a simple projection `is_distinct:= true`. It needs to be that first row is `true` and all subsequent rows are `false`. The overhead of using `MarkDistinctOperator` is building a hash set of one entry and checking everything against it is non-zero, but may not be high enough as well. Furthermore, if a constant value is propagated using run-length encoding, adding support for that to `MarkDistinctOperator` might be more useful anyway.

mbasmanova commented:
@sopel39 @martint Karol, Martin, is this good to go?

sopel39 commented:
For me it's OK. I think only @martint left.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:
```diff
@@ -366,8 +367,73 @@ public PlanWithProperties visitWindow(WindowNode node, StreamPreferredProperties
+            if (remainingSymbols.isEmpty()) {
+                // This happens when all distinct symbols are constants.
+                // In that case, keep the first symbol (don't drop them all).
+                return ImmutableList.of(node.getDistinctSymbols().get(0));
```
One minor improvement we could make: pick a BIGINT column (we have a dedicated optimization for that type. See BigintGroupByHash) or a "cheap" type if available (i.e., any of the fixed-width types that encodes to a primitive type)

mbasmanova replied:
@martint This would be a nice improvement. Since it is not a one-liner and requires additional testing, I'll plan to do this in a separate PR.
